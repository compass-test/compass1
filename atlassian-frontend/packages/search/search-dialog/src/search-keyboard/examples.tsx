import React, { FunctionComponent, useEffect } from 'react';
import {
  KeyboardHighlightProvider,
  useKeyboardNavigation,
} from './search-keyboard';
import faker from 'faker';

const Dummy: FunctionComponent<{}> = () => {
  const [isKeyboardHighlighted, ref] = useKeyboardNavigation<HTMLDivElement>();
  return (
    <div ref={ref}>{isKeyboardHighlighted ? 'Highlighted' : '---------'}</div>
  );
};

const DummyAsync: FunctionComponent<{}> = (props) => {
  const [loaded, setLoaded] = React.useState(false);

  useEffect(() => {
    setTimeout(
      () => setLoaded(true),
      faker.random.number({
        min: 300,
        max: 5000,
        precision: 1,
      }),
    );
  }, []);

  if (!loaded) {
    return <div>Loading</div>;
  } else {
    return <Dummy {...props} />;
  }
};

const keys = [...Array(20)].map((_, i) => i);

export const Basic = () => (
  <>
    <div>Default</div>
    {keys.map((key) => (
      <Dummy key={key} />
    ))}
  </>
);

export const ShuffledElements = () => (
  <>
    <div>Shuffled</div>
    <div>
      <p>Section 1</p>
      {keys.slice(0, 3).map((key) => (
        <DummyAsync key={key} />
      ))}
    </div>
    <div>
      <p>Section 2</p>
      {keys.slice(3, 6).map((key) => (
        <DummyAsync key={key} />
      ))}
    </div>
    {keys.slice(6, 9).map((key) => (
      <DummyAsync key={key} />
    ))}
  </>
);

export default {
  title: 'Search Dialog/Search Keyboard',
  decorators: [
    (story: () => React.ElementType) => (
      <KeyboardHighlightProvider listenerNode={document.body}>
        {story()}
      </KeyboardHighlightProvider>
    ),
  ],
};
