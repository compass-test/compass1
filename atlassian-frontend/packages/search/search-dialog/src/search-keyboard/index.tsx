export {
  useKeyboardNavigation,
  KeyboardHighlightProvider,
} from './search-keyboard';
export { TestSearchKeyboardProvider } from './search-keyboard-test-helper';
