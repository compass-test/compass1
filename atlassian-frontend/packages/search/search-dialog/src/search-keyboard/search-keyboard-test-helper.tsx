import React from 'react';
import {
  KeyboardHighlightContext,
  RegisterKeyboardEnabled,
} from './search-keyboard';

interface Props {
  allHighlighted: boolean;
}

export class TestSearchKeyboardProvider extends React.Component<Props> {
  timeout: ReturnType<typeof setTimeout> | undefined;

  componentWillUnmount() {
    this.timeout && clearTimeout(this.timeout);
  }

  registerKeyboardEnabled: RegisterKeyboardEnabled<any> = (
    _,
    __,
    onHighlight,
  ) => {
    this.timeout = setTimeout(() => onHighlight(this.props.allHighlighted), 0);
    return -1;
  };

  render() {
    return (
      <KeyboardHighlightContext.Provider
        value={{
          registerKeyboardEnabled: this.registerKeyboardEnabled,
          unregisterKeyboardEnabled: () => {},
        }}
      >
        {this.props.children}
      </KeyboardHighlightContext.Provider>
    );
  }
}
