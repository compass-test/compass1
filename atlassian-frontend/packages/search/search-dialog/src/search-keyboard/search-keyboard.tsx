import React, { useContext, useEffect, useCallback } from 'react';

type OnHighlightChanged = (isHighlighted: boolean) => void;
type OnKeyDown = (e: KeyboardEvent, target: HTMLElement) => void;
export type RegisterKeyboardEnabled<U extends HTMLElement> = (
  currentId: number | null,
  domRef: U,
  onHighlightChanged: OnHighlightChanged,
  onKeyDownWhenHighlighted: OnKeyDown | undefined,
) => number;

interface KeyboardContextProps<U extends HTMLElement = HTMLElement> {
  registerKeyboardEnabled?: RegisterKeyboardEnabled<U>;
  unregisterKeyboardEnabled?: (ref: number) => void;
}

export const KeyboardHighlightContext = React.createContext<
  KeyboardContextProps
>({
  registerKeyboardEnabled: undefined,
  unregisterKeyboardEnabled: undefined,
});

export interface KeyboardNavigationHookOptions {
  onKeydownCallback: OnKeyDown;
}

export const useKeyboardNavigation = <U extends HTMLElement>(
  options?: KeyboardNavigationHookOptions,
): [boolean, (ref: U) => void] => {
  const [ref, setRef] = React.useState<U | null>(null);
  const registrationId = React.useRef<number | null>(null);
  const { registerKeyboardEnabled, unregisterKeyboardEnabled } = useContext(
    KeyboardHighlightContext,
  );
  const [isHighlighted, setIsHighlighted] = React.useState(false);
  const onKeydownCallback = options?.onKeydownCallback;

  const unregister = useCallback(() => {
    if (registrationId.current && unregisterKeyboardEnabled) {
      // If the ref got removed or changed we unregister the current ref
      unregisterKeyboardEnabled(registrationId.current);
    }
  }, [unregisterKeyboardEnabled]);

  // Will unmount effect
  useEffect(() => () => unregister(), [unregister]);

  // Ref change to null
  useEffect(() => {
    !ref && unregister();
  }, [ref, unregister]);

  // Everything else
  useEffect(() => {
    if (!registerKeyboardEnabled || !unregisterKeyboardEnabled) {
      return;
    }

    if (!ref) {
      return;
    }

    // Then we register the new ref
    if (ref) {
      // Another check here to ensure that the ref we get is actually a DOM node, there's probably better
      // ways to make this check but it doesn't matter too much.
      // If what we get is NOT a DOM node we will log a warning and default back to use React.findDOMNode
      if (
        !ref.compareDocumentPosition ||
        !ref.addEventListener ||
        !ref.removeEventListener
      ) {
        // eslint-disable-next-line
        console.error(
          'It is expected that any keyboard highlighting enabled component also passes the ref all the way to the DOM node.',
          'It appears that the `ref` is not fowarded all the way to the DOM node, please use `React.forwardRef to ensure this is so`',
          'Falling back to using ReactDOM.findDOMNode, this is not recommended and this fallback may be removed in the future',
        );

        return;
      }

      registrationId.current = registerKeyboardEnabled(
        registrationId.current,
        ref,
        setIsHighlighted,
        onKeydownCallback,
      );
    }
  }, [
    ref,
    setIsHighlighted,
    onKeydownCallback,
    registerKeyboardEnabled,
    unregisterKeyboardEnabled,
  ]);

  return [isHighlighted, setRef];
};

export interface KeyboardHighlightProviderProps {
  listenerNode: HTMLElement | undefined | null;
}

export class KeyboardHighlightProvider extends React.Component<
  KeyboardHighlightProviderProps
> {
  private currentlySelected: number | undefined = undefined;

  private keyboardNavComponents: {
    id: number;
    domRef: HTMLElement;
    onHighlightChanged: OnHighlightChanged;
    onKeyDownWhenHighlighted: OnKeyDown | undefined;
  }[] = [];

  private registerKeyboardEnabled = (
    currentId: number | null,
    domRef: HTMLElement,
    onHighlightChanged: OnHighlightChanged,
    onKeyDownWhenHighlighted: OnKeyDown | undefined,
  ) => {
    const indexAt = this.keyboardNavComponents.findIndex(
      ({ id }) => id === currentId,
    );
    // If the currentId is null or we can't find any records of the id we will create a new entry
    const isNew: boolean =
      currentId === null || currentId === undefined || indexAt === -1;
    const id: number = isNew
      ? Math.max(
          ...this.keyboardNavComponents.map((component) => component.id),
          0,
        ) + 1
      : this.keyboardNavComponents[indexAt].id;

    if (isNew) {
      this.keyboardNavComponents.push({
        id,
        domRef,
        onHighlightChanged,
        onKeyDownWhenHighlighted,
      });
    } else {
      // Otherwise we will update the current entry
      this.keyboardNavComponents[indexAt] = {
        id,
        domRef,
        onHighlightChanged,
        onKeyDownWhenHighlighted,
      };
    }

    this.keyboardNavComponents = this.keyboardNavComponents.sort((c1, c2) => {
      const result = c1.domRef.compareDocumentPosition(c2.domRef);
      if (
        result & Node.DOCUMENT_POSITION_FOLLOWING ||
        result & Node.DOCUMENT_POSITION_CONTAINED_BY
      ) {
        return -1;
      } else if (
        result & Node.DOCUMENT_POSITION_PRECEDING ||
        result & Node.DOCUMENT_POSITION_CONTAINS
      ) {
        return 1;
      }

      return 0;
    });

    return id;
  };

  private unregisterKeyboardEnabled = (unregisterRef: number) => {
    if (this.currentlySelected === unregisterRef) {
      this.currentlySelected = undefined;
    }

    this.keyboardNavComponents = this.keyboardNavComponents.filter(
      ({ id: ref }) => ref !== unregisterRef,
    );
  };

  onKeyDownCallbackAt = (index: number | undefined, e: KeyboardEvent) => {
    if (
      typeof index !== 'undefined' &&
      index >= 0 &&
      index < this.keyboardNavComponents.length
    ) {
      const onKeyDown = this.keyboardNavComponents[index]
        .onKeyDownWhenHighlighted;
      onKeyDown && onKeyDown(e, this.keyboardNavComponents[index].domRef);
    }
  };

  unhighlightComponentAt = (index: number | undefined) => {
    if (
      typeof index !== 'undefined' &&
      index >= 0 &&
      index < this.keyboardNavComponents.length
    ) {
      this.keyboardNavComponents[index].onHighlightChanged(false);
    }
  };

  highlightComponentAt = (index: number) => {
    if (index >= 0 && index < this.keyboardNavComponents.length) {
      this.currentlySelected = this.keyboardNavComponents[index].id;
      this.keyboardNavComponents[index].onHighlightChanged(true);
    }
  };

  getSelectedIndex = () => {
    if (!this.currentlySelected) {
      return undefined;
    }

    return this.keyboardNavComponents.findIndex(
      ({ id }) => id === this.currentlySelected,
    );
  };

  onKeyDown = (e: KeyboardEvent) => {
    if (this.keyboardNavComponents.length === 0) {
      return;
    }

    if (e.isComposing || e.keyCode === 229) {
      return;
    }

    let currentIndex = this.getSelectedIndex();
    const prevIndex = currentIndex;

    if (e.code === 'ArrowDown' || e.keyCode === 40) {
      if (!e.shiftKey) {
        if (typeof currentIndex === 'undefined') {
          currentIndex = 0;
        } else {
          currentIndex = (currentIndex + 1) % this.keyboardNavComponents.length;
        }
        e.preventDefault();
      }
    } else if (e.code === 'ArrowUp' || e.keyCode === 38) {
      if (!e.shiftKey) {
        if (typeof currentIndex === 'undefined') {
          currentIndex = this.keyboardNavComponents.length - 1;
        } else {
          currentIndex -= 1;

          if (currentIndex < 0) {
            currentIndex =
              this.keyboardNavComponents.length - Math.abs(currentIndex);
          }
        }
        e.preventDefault();
      }
    } else {
      if (typeof currentIndex !== 'undefined') {
        this.onKeyDownCallbackAt(currentIndex, e);
      }
      return;
    }

    if (prevIndex !== currentIndex && currentIndex !== undefined) {
      this.unhighlightComponentAt(prevIndex);
      this.highlightComponentAt(currentIndex);
    }
  };

  attach = (listenerNode: HTMLElement | undefined | null) => {
    if (listenerNode) {
      listenerNode.addEventListener('keydown', this.onKeyDown);
    }
  };

  detach = (listenerNode: HTMLElement | undefined | null) => {
    if (listenerNode) {
      listenerNode.removeEventListener('keydown', this.onKeyDown);
    }
  };

  componentDidMount() {
    this.attach(this.props.listenerNode);
  }

  componentDidUpdate(prevProps: KeyboardHighlightProviderProps) {
    if (prevProps.listenerNode !== this.props.listenerNode) {
      this.detach(prevProps.listenerNode);
      this.attach(this.props.listenerNode);
    }
  }

  componentWillUnmount() {
    this.detach(this.props.listenerNode);
  }

  render() {
    return (
      <KeyboardHighlightContext.Provider
        value={{
          registerKeyboardEnabled: this.registerKeyboardEnabled,
          unregisterKeyboardEnabled: this.unregisterKeyboardEnabled,
        }}
      >
        {this.props.children}
      </KeyboardHighlightContext.Provider>
    );
  }
}
