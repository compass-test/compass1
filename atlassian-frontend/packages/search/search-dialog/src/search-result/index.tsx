import { Props } from './search-result';

export { SearchResult } from './search-result';

export type SearchResultProps = Props;
