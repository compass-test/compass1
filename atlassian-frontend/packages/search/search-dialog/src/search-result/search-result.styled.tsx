import styled from '@emotion/styled';
import {
  grid,
  primaryTextColour,
  visitedLinkColor,
  itemHighlightColor,
  itemLinkSecondaryColor,
  dialogPageTakeoverWidth,
  nowrapEllipsis,
} from '../style-utils';
import { Link, LinkComponentProps } from '../search-link-component';

interface ResultLinkProps {
  isCollapsed?: boolean;
  isKeyboardHighlighted?: boolean;
}

export const ResultLink = styled(Link)<LinkComponentProps & ResultLinkProps>`
  display: flex;
  height: ${(props) =>
    props.isCollapsed ? grid.multiple(5).px : grid.multiple(4).px};
  align-items: center;
  flex-flow: row nowrap;
  position: relative;

  background-color: ${(props) =>
    props.isKeyboardHighlighted ? itemHighlightColor : 'transparent'};
  color: ${primaryTextColour};
  text-decoration: none;

  padding: 0 ${grid.twice().px};

  &:hover {
    background-color: ${itemHighlightColor};
    color: ${primaryTextColour};
    text-decoration: none;
  }

  &:focus {
    outline-offset: -2px;
  }

  &:visited {
    color: ${visitedLinkColor};
  }
`;

export const ReturnIconContainer = styled.div<{ isCollapsed?: boolean }>`
  position: absolute;
  border-radius: 3px;
  margin: ${grid.half().px};
  width: ${grid.multiple(3).px};
  height: ${grid.twice().px};
  right: ${grid.half().px};
  top: ${(props) => (props && props.isCollapsed ? grid.px : grid.half().px)};
`;

export const IconContainer = styled.div`
  width: ${grid.twice().px};
  margin-right: ${grid.twice().px};
  flex: 0 0 ${grid.twice().px};
  line-height: ${grid.twice().px};
  display: flex;
  align-items: center;
`;

export const ResultTitleContainer = styled.div<{ isCollapsed?: boolean }>`
  ${nowrapEllipsis}

  width: ${(props) =>
    props.isCollapsed ? grid.multiple(52).px : grid.multiple(47).px};
  max-width: ${grid.multiple(60).px};
  line-height: 16px;
  flex: 1 0 auto;

  @media screen and (max-width: ${dialogPageTakeoverWidth}px) {
    width: 100%;
  }
`;

export const ResultLabel = styled.span`
  ${nowrapEllipsis}

  color: #5E6C84;
  font-size: 14px;
  line-height: 16px;
  margin-right: ${grid.px};
`;

export const ResultDetail = styled.div`
  ${nowrapEllipsis}

  font-size: 11px;
  line-height: 16px;
  color: ${itemLinkSecondaryColor};
  margin-left: ${grid.multiple(3).px};
  width: ${grid.multiple(19).px};
`;

export const CollapsedResultDetail = styled.div`
  ${nowrapEllipsis}

  font-size: 11px;
  line-height: 14px;
  color: ${itemLinkSecondaryColor};
  max-width: ${grid.multiple(19).px};

  @media screen and (max-width: ${dialogPageTakeoverWidth}px) {
    max-width: initial;
  }
`;

export const CollapsedResultDetailWrapper = styled.div`
  display: flex;
`;

export const DetailsSeparator = styled.div`
  font-size: 11px;
  line-height: 14px;
  color: ${itemLinkSecondaryColor};
  padding: 0 ${grid.px};
`;
