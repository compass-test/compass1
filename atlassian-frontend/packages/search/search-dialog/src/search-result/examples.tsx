import React from 'react';
import { action } from '@storybook/addon-actions';
import styled from '@emotion/styled';
import BoardIcon from '@atlaskit/icon/glyph/board';
import ListIcon from '@atlaskit/icon/glyph/list';
import FilterIcon from '@atlaskit/icon/glyph/filter';
import { SearchResult } from './search-result';
import { dialogWidth } from '../style-utils';
import { TestSearchKeyboardProvider } from '../search-keyboard';

// Fake data
// -----------------------------
const issueProps = {
  containerDetail: 'This is the container detail',
  href: '#',
  icon: <img src="http://lorempixel.com/24/24/cats" alt="business" />,
  label: 'GP-103',
  timeDetail: <div>Updated May 18, 2019</div>,
  title: 'Query parameters missing from landing page URL',
};

const boardProps = {
  containerDetail: 'e-services primary Lead',
  href:
    '/secure/RapidBoard.jspa?rapidView=da417a6a-07e0-4526-80c8-27d70b583b00&searchSessionId=6ab74e81-c430-40aa-ba85-80302e9d0280',
  icon: <BoardIcon label="board" />,
  title: 'Board 393 example',
};

const filterProps = {
  containerDetail: 'e-services primary Lead',
  href:
    '/secure/RapidBoard.jspa?rapidView=da417a6a-07e0-4526-80c8-27d70b583b00&searchSessionId=6ab74e81-c430-40aa-ba85-80302e9d0280',
  icon: <FilterIcon label="filter" />,
  title: 'Filtering is a good thing',
};

const projectProps = {
  href: '/browse/CIIAA?searchSessionId=979f5747-def7-4af1-91d9-7173564db14d',
  icon: <img src="http://lorempixel.com/24/24/cats" alt="business" />,
  title: 'harum illo quos minima facilis',
};

const defaultProps = {
  containerDetail: 'Facilities Workplace Support',
  href: 'https://www.atlassian.com/',
  icon: <ListIcon label="I'm an icon" />,
  onClick: (e: any) => {
    e.preventDefault();
    action('onClick')(e);
  },
  timeDetail: <div>Updated May 14, 2019</div>,
  title:
    'PSA: 752 feature flags got evaluated to default value specified in the code! :omg:',
};

// Stories
// -----------------------------
export const Basic = () => (
  <SearchResult
    {...defaultProps}
    href={`https://www.atlassian.com/${Math.random()}`}
  />
);

export const Collapsed = () => <SearchResult {...defaultProps} isCollapsed />;

export const WithVisitedLink = () => (
  <SearchResult {...defaultProps} href={`https://www.atlassian.com/`} />
);

export const KeyboardHighlighted = () => (
  <TestSearchKeyboardProvider allHighlighted>
    <SearchResult {...defaultProps} />
  </TestSearchKeyboardProvider>
);

export const MissingSomeDetail = () => (
  <>
    <p>Missing time</p>
    <SearchResult
      {...defaultProps}
      timeDetail={undefined}
      href={`https://www.atlassian.com/`}
    />
    <p>Missing container</p>
    <SearchResult
      {...defaultProps}
      containerDetail={undefined}
      href={`https://www.atlassian.com/`}
    />
    <p>Missing both</p>
    <SearchResult
      {...defaultProps}
      timeDetail={undefined}
      containerDetail={undefined}
      href={`https://www.atlassian.com/`}
    />
  </>
);

export const WithLongDetail = () => (
  <SearchResult
    {...defaultProps}
    containerDetail={'Facilities Workplace Support For Technical Staff'}
    timeDetail={<div>Updated yesterday at 11:04 AM</div>}
    href={`https://www.atlassian.com/`}
  />
);

export const InDialogSizedContainers = () => {
  const DialogSizedContainer = styled.div`
    width: ${dialogWidth};
    border: 1px solid grey;
    margin-bottom: 8px;
    padding: 8px;
    box-sizing: border-box;
  `;

  return (
    <DialogSizedContainer>
      <SearchResult
        {...defaultProps}
        href={`https://www.atlassian.com/${Math.random()}`}
      />
      <TestSearchKeyboardProvider allHighlighted>
        <SearchResult {...defaultProps} />
      </TestSearchKeyboardProvider>
    </DialogSizedContainer>
  );
};

export const SomeWithLabelBeforeTitle = () => (
  <>
    <SearchResult
      {...defaultProps}
      href={`https://www.atlassian.com/${Math.random()}`}
      label="QS-1191"
    />
    <SearchResult
      {...defaultProps}
      href={`https://www.atlassian.com/${Math.random()}`}
      label="QS-349"
    />
    <SearchResult
      {...defaultProps}
      href={`https://www.atlassian.com/${Math.random()}`}
      label="QS-55391"
    />
    <SearchResult
      {...defaultProps}
      href={`https://www.atlassian.com/${Math.random()}`}
    />
  </>
);

export const WithCustomLinkComponent = () => (
  <SearchResult
    {...defaultProps}
    href={`https://www.atlassian.com/${Math.random()}`}
    linkComponent={styled.div``}
  />
);

export const IssueCollapsed = () => (
  <SearchResult {...issueProps} isCollapsed />
);

export const IssueCollapsedLongTitle = () => (
  <SearchResult
    {...issueProps}
    title="This is a really, really long title that should test how resilient a design is!"
    isCollapsed
  />
);

export const BoardCollapsed = () => (
  <SearchResult {...boardProps} isCollapsed />
);

export const BoardCollapsedLongTitle = () => (
  <SearchResult
    {...boardProps}
    title="This is a really, really long title that should test how resilient a design is!"
    isCollapsed
  />
);

export const ProjectCollapsed = () => (
  <SearchResult {...projectProps} isCollapsed />
);

export const ProjectCollapsedLongTitle = () => (
  <SearchResult
    {...projectProps}
    title="This is a really, really long title that should test how resilient a design is!"
    isCollapsed
  />
);

export const FlterCollapsed = () => (
  <SearchResult {...filterProps} isCollapsed />
);

export const FilterCollapsedLongTitle = () => (
  <SearchResult
    {...filterProps}
    title="This is a really, really long title that should test how resilient a design is!"
    isCollapsed
  />
);

export default {
  title: 'Search Dialog/Search Result',
  decorators: [
    (story: () => React.ElementType) => (
      <TestSearchKeyboardProvider allHighlighted={false}>
        {story()}
      </TestSearchKeyboardProvider>
    ),
  ],
};
