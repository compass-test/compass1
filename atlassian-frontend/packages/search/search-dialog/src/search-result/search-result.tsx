import React, { ReactNode, useRef, useEffect } from 'react';
import {
  ResultTitleContainer,
  ResultLabel,
  ResultDetail,
  IconContainer,
  ResultLink,
  CollapsedResultDetail,
  DetailsSeparator,
  CollapsedResultDetailWrapper,
  ReturnIconContainer,
} from './search-result.styled';
import { useKeyboardNavigation } from '../search-keyboard';
import { ReturnIcon } from '../icons';
import { LinkComponent } from '../search-link-component';

// todo - set a default value for "icon" and change to required
export interface Props {
  /**
   * Link the result
   */
  href: string;
  /**
   * Display text (and tooltip) for the result item.
   */
  title: string;
  /**
   * Toggle for whether to show the search results in a collapsed mode.
   * The collapsed mode is used when width is restricted (e.g. mobile view). This mode will take up more vertical width in exchange for less horizontal width.
   */
  isCollapsed?: boolean;
  /**
   * This will replace all links with the given component. This will be used to ensure that results provided will be properly SPA transitioned.
   * Do not use this for styling.
   */
  linkComponent?: LinkComponent;
  /**
   * An icon for the search result.
   */
  icon?: JSX.Element;
  /**
   * An label that will be shown before the title. You can use this to display the id of the item, e.g. issue key.
   */
  label?: ReactNode;
  /**
   * Display text (and tooltip) for the result item container.
   */
  containerDetail?: string;
  /**
   * Displays any relevant date time information of the result item, e.g. last updated time.
   */
  timeDetail?: JSX.Element;
  /**
   * Callback that is called when the result item is either left clicked or the user pressed enter while focusing on the element.
   */
  onSelect?: (e: React.MouseEvent | KeyboardEvent) => void;
  /**
   * Callback that is called when the element is highlighted via keyboard navigation (this will not trigger on focus).
   */
  onHighlighted?: () => void;
}

const onKeyDown = (e: KeyboardEvent, target: HTMLElement) => {
  if (e.key === 'Enter' && !e.shiftKey) {
    e.preventDefault();

    // The ref here is the wrapper div however we simulate a click on the `ResultLinkComponent` that it wraps.
    (target.firstElementChild as HTMLElement)?.click?.();
  }
};

export const SearchResult: React.FunctionComponent<Props> = ({
  href,
  linkComponent,
  icon,
  label,
  title,
  containerDetail,
  timeDetail,
  onSelect,
  isCollapsed,
  onHighlighted,
  ...rest
}) => {
  const [isKeyboardHighlighted, ref] = useKeyboardNavigation<HTMLDivElement>({
    onKeydownCallback: onKeyDown,
  });

  const previousHighlight = useRef(false);

  // Callback for when the component is highlighted
  useEffect(() => {
    if (!previousHighlight.current && isKeyboardHighlighted) {
      onHighlighted?.();
    }

    previousHighlight.current = isKeyboardHighlighted;
  }, [isKeyboardHighlighted, previousHighlight, onHighlighted]);

  const getResultContent = () => (
    <>
      {isCollapsed ? (
        <div style={{ display: 'block', minWidth: 0 }}>
          <ResultTitleContainer isCollapsed>
            {typeof label === 'string' ? (
              <ResultLabel title={title}>{label}</ResultLabel>
            ) : (
              label
            )}
            <span title={title}>{title}</span>
          </ResultTitleContainer>
          <CollapsedResultDetailWrapper>
            <CollapsedResultDetail title={containerDetail}>
              {containerDetail}
            </CollapsedResultDetail>
            {timeDetail && (
              <>
                <DetailsSeparator>{'â€¢'}</DetailsSeparator>
                <CollapsedResultDetail>{timeDetail}</CollapsedResultDetail>
              </>
            )}
          </CollapsedResultDetailWrapper>
        </div>
      ) : (
        <>
          <ResultTitleContainer>
            {typeof label === 'string' ? (
              <ResultLabel title={title}>{label}</ResultLabel>
            ) : (
              label
            )}
            <span title={title}>{title}</span>
          </ResultTitleContainer>
          <ResultDetail title={containerDetail}>{containerDetail}</ResultDetail>
          {timeDetail ? <ResultDetail>{timeDetail}</ResultDetail> : null}
        </>
      )}
    </>
  );

  return (
    <div ref={ref} onClick={onSelect} role="none">
      <ResultLink
        linkComponent={linkComponent}
        href={href}
        isKeyboardHighlighted={isKeyboardHighlighted}
        isCollapsed={isCollapsed}
        {...rest}
      >
        <IconContainer>{icon}</IconContainer>
        {getResultContent()}
        {isKeyboardHighlighted ? (
          <ReturnIconContainer isCollapsed={isCollapsed}>
            <ReturnIcon inverted />
          </ReturnIconContainer>
        ) : null}
      </ResultLink>
    </div>
  );
};
