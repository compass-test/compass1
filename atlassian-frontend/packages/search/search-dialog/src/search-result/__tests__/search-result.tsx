import React, { ComponentProps, PropsWithChildren } from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { SearchResult } from '../search-result';
import { ResultTitleContainer } from '../search-result.styled';
import { useKeyboardNavigation } from '../../search-keyboard';

jest.mock('../search-result.styled', () => {
  return {
    ResultTitleContainer: ({ children }: PropsWithChildren<{}>) => children,
    ResultLabel: ({ children }: PropsWithChildren<{}>) => children,
    ResultDetail: () => 'div',
    IconContainer: () => 'div',
    ReturnIconContainer: () => 'div',
    ResultLink: () => 'div',
  };
});

jest.mock('../../search-keyboard', () => ({
  useKeyboardNavigation: jest.fn(),
}));

describe('<SearchResult />', () => {
  const updateWrapper = (wrapper: ReactWrapper) => {
    wrapper
      .setProps({
        someProp: 'changed',
      })
      .update();
  };

  const DummyLinkComponent: React.FunctionComponent = ({
    children,
    ...rest
  }) => <a {...rest}>{children}</a>;

  beforeEach(() => {
    jest.clearAllMocks();
    (useKeyboardNavigation as jest.Mock).mockReturnValue([false]);
  });

  const defaultProps: ComponentProps<typeof SearchResult> = {
    href: 'www.atlassian.com',
    title: 'dinosaurs',
    containerDetail: 'jurassic period',
    timeDetail: <div>last updated 145m years ago</div>,
    icon: <div>Icon</div>,
  };

  it('SearchResult matches snapshot', () => {
    const wrapper = mount(<SearchResult {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('SearchResult matches snapshot when highlighted', () => {
    (useKeyboardNavigation as jest.Mock).mockReturnValue([true]);
    const wrapper = mount(<SearchResult {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('SearchResult matches snapshot with custom component', () => {
    const wrapper = mount(
      <SearchResult {...defaultProps} linkComponent={DummyLinkComponent} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('SearchResult generates a label prefix when supplied', () => {
    const wrapper = mount(<SearchResult {...defaultProps} label="Hello" />);
    expect(wrapper.find(ResultTitleContainer)).toMatchSnapshot();
  });

  it('onSelect is called when a search result is clicked', () => {
    const onSelect = jest.fn();
    const wrapper = mount(
      <SearchResult
        {...defaultProps}
        linkComponent={DummyLinkComponent}
        onSelect={onSelect}
      />,
    );

    wrapper.simulate('click');
    expect(onSelect).toHaveBeenCalled();
  });

  it('SearchResult calls onHighlighted when highlighted', () => {
    const onHighlighted = jest.fn();

    (useKeyboardNavigation as jest.Mock).mockReturnValue([false]);
    const wrapper = mount(
      <SearchResult {...defaultProps} onHighlighted={onHighlighted} />,
    );

    expect(onHighlighted).toBeCalledTimes(0);
    (useKeyboardNavigation as jest.Mock).mockReturnValue([true]);
    updateWrapper(wrapper);

    expect(onHighlighted).toBeCalledTimes(1);
  });

  it('SearchResult calls onHighlighted when highlighted on mount', () => {
    const onHighlighted = jest.fn();

    (useKeyboardNavigation as jest.Mock).mockReturnValue([true]);
    mount(<SearchResult {...defaultProps} onHighlighted={onHighlighted} />);

    expect(onHighlighted).toBeCalledTimes(1);
  });

  it('SearchResult does not call onHighlighted when unhighlighted', () => {
    const onHighlighted = jest.fn();

    (useKeyboardNavigation as jest.Mock).mockReturnValue([true]);
    const wrapper = mount(
      <SearchResult {...defaultProps} onHighlighted={onHighlighted} />,
    );

    onHighlighted.mockReset();
    (useKeyboardNavigation as jest.Mock).mockReturnValue([false]);
    updateWrapper(wrapper);

    expect(onHighlighted).toBeCalledTimes(0);
  });

  it('SearchResult does not call onHighlighted when highlight does not change', () => {
    const onHighlighted = jest.fn();

    (useKeyboardNavigation as jest.Mock).mockReturnValue([true]);
    const wrapper = mount(
      <SearchResult {...defaultProps} onHighlighted={onHighlighted} />,
    );

    onHighlighted.mockReset();
    (useKeyboardNavigation as jest.Mock).mockReturnValue([true]);
    updateWrapper(wrapper);

    expect(onHighlighted).toBeCalledTimes(0);
  });
});
