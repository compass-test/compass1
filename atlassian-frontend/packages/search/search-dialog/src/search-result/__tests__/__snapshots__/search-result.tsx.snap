// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResult /> SearchResult generates a label prefix when supplied 1`] = `null`;

exports[`<SearchResult /> SearchResult matches snapshot 1`] = `
<SearchResult
  containerDetail="jurassic period"
  href="www.atlassian.com"
  icon={
    <div>
      Icon
    </div>
  }
  timeDetail={
    <div>
      last updated 145m years ago
    </div>
  }
  title="dinosaurs"
>
  <div
    role="none"
  >
    <ResultLink
      href="www.atlassian.com"
      isKeyboardHighlighted={false}
    >
      div
    </ResultLink>
  </div>
</SearchResult>
`;

exports[`<SearchResult /> SearchResult matches snapshot when highlighted 1`] = `
<SearchResult
  containerDetail="jurassic period"
  href="www.atlassian.com"
  icon={
    <div>
      Icon
    </div>
  }
  timeDetail={
    <div>
      last updated 145m years ago
    </div>
  }
  title="dinosaurs"
>
  <div
    role="none"
  >
    <ResultLink
      href="www.atlassian.com"
      isKeyboardHighlighted={true}
    >
      div
    </ResultLink>
  </div>
</SearchResult>
`;

exports[`<SearchResult /> SearchResult matches snapshot with custom component 1`] = `
<SearchResult
  containerDetail="jurassic period"
  href="www.atlassian.com"
  icon={
    <div>
      Icon
    </div>
  }
  linkComponent={[Function]}
  timeDetail={
    <div>
      last updated 145m years ago
    </div>
  }
  title="dinosaurs"
>
  <div
    role="none"
  >
    <ResultLink
      href="www.atlassian.com"
      isKeyboardHighlighted={false}
      linkComponent={[Function]}
    >
      div
    </ResultLink>
  </div>
</SearchResult>
`;
